crates/daphne-server/src/storage_proxy_connection/kv/request_coalescer.rs

```diff
@@ -0,0 +1,494 @@
+        }
+
+        self.inflight.lock().unwrap().clear();
+        self.shutting_down
+            .store(false, std::sync::atomic::Ordering::SeqCst);
+    }
+
+    pub async fn coalesce<F, Fut, R, E>(
+        &self,
+        key: String,
+        executor: F,
+    ) -> Result<Option<Marc<R>>, Marc<E>>
+    where
+        R: Send + Sync + 'static,
+        E: Send + Sync + 'static,
+        F: FnOnce() -> Fut,
+        Fut: Future<Output = Result<Option<Marc<R>>, E>>,
+    {
+        let result = loop {
+            #[cfg(feature = "test-utils")]
+            assert!(
+                !self.shutting_down.load(std::sync::atomic::Ordering::SeqCst),
+                "RequestCoalescer is shutting down"
+            );
+
+            let op_key = OperationKey {
+                op_id: operation_id::of::<F, Fut, R, E>(),
+                key: key.clone(),
+            };
+            let role = match self.inflight.lock().unwrap().entry(op_key.clone()) {
+                Entry::Occupied(receiver) => {
+                    CoalescingRole::Wait(ResponseReceiver::new(receiver.get().clone()))
+                }
+                Entry::Vacant(slot) => {
+                    let (tx, rx) = watch::channel(None);
+                    slot.insert(rx);
+                    CoalescingRole::Execute(ResponseSender::new(tx, op_key, self))
+                }
+            };
+            break match role {
+                CoalescingRole::Wait(receiver) => match receiver.wait().await {
+                    Ok(v) => Ok(v),
+                    Err(CoalesceError::Canceled) => continue,
+                    Err(CoalesceError::Error(e)) => Err(e),
+                },
+                CoalescingRole::Execute(sender) => {
+                    let result = executor().await.map_err(Marc::new);
+                    sender.send(result.clone());
+                    result
+                }
+            };
+        };
+        result
+    }
+}
+
+/// An error that can occur while executing a coalesced request.
+enum CoalesceError<E: ?Sized + 'static> {
+    /// The `executor` task that is executing has been cancelled and the waiting tasks that receive
+    /// this error should retry the operation.
+    Canceled,
+    /// An error that occurred during execution of the coalesced operation.
+    ///
+    /// This error is behind an [`Arc`] because it will be shared between all `waiters`. This is
+    /// needed to avoid useless deep [`Clone`]s but also because most errors don't implement
+    /// [`Clone`].
+    Error(Marc<E>),
+}
+
+impl<E: ?Sized> Clone for CoalesceError<E> {
+    fn clone(&self) -> Self {
+        match self {
+            Self::Canceled => Self::Canceled,
+            Self::Error(e) => Self::Error(e.clone()),
+        }
+    }
+}
+
+type CoalescedValue =
+    Result<Option<Marc<dyn Any + Send + Sync>>, CoalesceError<dyn Any + Send + Sync>>;
+
+/// The send half to be used by the `executor` to send the result of it's computation to the
```

Comment:
```suggestion
/// The send have to be used by the `executor` to send the result of it's computation to the
```

Reply:
This is actually correct. A channel has two parts, the sending part and the receiving part, commonly referred to as the sending and receiving halves